pkg/workflow/workflow.go

```diff
@@ -0,0 +1,232 @@
+package workflow
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pborman/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2/klogr"
+	diskv1beta1 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta1"
+)
+
+type workflowKey struct{}
+
+type Workflow struct {
+	name           string
+	errSet         *sync.Map
+	startTime      time.Time
+	requestID      string
+	logger         Logger
+	parentWorkflow *Workflow
+	numWait        int32 // numWait should always be number of children + 1(itself)
+}
+
+type Option func(*Workflow)
+
+func WithDetails(details ...interface{}) Option {
+	return func(w *Workflow) {
+		w.logger = Logger{w.logger.WithValues(details...)}
+	}
+}
+
+func GetWorkflowFromObj(ctx context.Context, obj k8sRuntime.Object, details ...interface{}) (context.Context, Workflow) {
+	requestID := GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	var startTime time.Time
+	if startTimeStr != "" {
+		startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	w := Workflow{
+		requestID: requestID,
+		startTime: startTime,
+		logger:    Logger{klogr.New().WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, requestID)...)},
+	}
+	w.parentWorkflow = &w
+	return w.SaveToContext(ctx), w
+}
+
+func GetObjectDetails(obj k8sRuntime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+// New starts a new child workflow if there is already a workflow saved to the context otherwise a new workflow
+func New(ctx context.Context, opts ...Option) (context.Context, Workflow) {
```

Comment:
When using the Functional Options Pattern, I like to structure my `New` function so that options are processed first and default values are applied afterward. This way, you avoid duplicating work if an option overrides a value. You can look at my implementation of [`batch.NewProcessor`](https://github.com/edreed/cloud-provider-azure/blob/817d5114fc89f8b9e8204c4138a8c4f42b093d7d/pkg/batch/batch.go#L350) as an example.